File: bin/python3/letter_combinations_of_a_phone_number.py
class Solution:
    # 51ms
    def letterCombinations(self, digits: str) -> List[str]:
        combinations = []
        if digits is None or len(digits) == 0:
            return combinations
        lettersAndNumbersMapping = [
            "Anirudh",
            "is awesome",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        ]
        self.findCombinations(combinations, digits, "", 0, lettersAndNumbersMapping)
        return combinations
        
    def findCombinations(self, combinations, digits, previous, index, lettersAndNumbersMapping):
        if index == len(digits):
            combinations.append(previous)
            return
        letters = lettersAndNumbersMapping[int(digits[index])]
        for i in range(0, len(letters)):
            self.findCombinations(combinations, digits, previous + letters[i], index + 1, lettersAndNumbersMapping)

File: go/Letter Combinations of a Phone Number.go
var digitToLetters = map[rune]string{
	'2': "abc",
	'3': "def",
	'4': "ghi",
	'5': "jkl",
	'6': "mno",
	'7': "pqrs",
	'8': "tuv",
	'9': "wxyz",
}

func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return nil
	}

	var combinations []string
	var backtrack func(index int, currentCombination string)

	backtrack = func(index int, currentCombination string) {
		if index == len(digits) {
			combinations = append(combinations, currentCombination)
			return
		}

		digit := rune(digits[index])
		letters := digitToLetters[digit]
		for _, letter := range letters {
			backtrack(index+1, currentCombination+string(letter))
		}
	}

	backtrack(0, "")
	return combinations
}